KAFKA-14551 Move/Rewrite LeaderEpochFileCache and its dependencies to the storage module.

satishd opened the pull request:
KAFKA-14551 Move/Rewrite LeaderEpochFileCache and its dependencies to the storage module.

For broader context on this change, you may want to look at KAFKA-14470: Move log layer to the storage module

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
KAFKA-14551 Move LeaderEpochCheckpointFile and its dependencies to the storage module.

satishd committed:
KAFKA-14551 Move LeaderEpochFileCache to the storage module.

ijuma commented on core/src/main/scala/kafka/server/checkpoints/OffsetCheckpointFile.scala:
```diff
@@ -62,12 +62,17 @@ trait OffsetCheckpoint {
  */
 class OffsetCheckpointFile(val file: File, logDirFailureChannel: LogDirFailureChannel = null) {
-  val checkpoint = new CheckpointFileWithFailureHandler[(TopicPartition, Long)](file, OffsetCheckpointFile.CurrentVersion,
+  val checkpoint = new CheckpointFileWithFailureHandler[Tuple2[TopicPartition, Long]](file, OffsetCheckpointFile.CurrentVersion,
```
What is the reason for this change?

satishd replied:
Nice catch! This is not needed, it seemed to be part of earlier refactoring.

ijuma commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -1286,31 +1286,31 @@ class UnifiedLog(@volatile var logStartOffset: Long,
         // it may not be true following a message format version bump as the epoch will not be available for
         // log entries written in the older format.
-        val earliestEpochEntry = leaderEpochCache.flatMap(_.earliestEntry)
+        val earliestEpochEntry = leaderEpochCache.flatMap(_.earliestEntry().asScala)
```
To avoid multiple conversions, I think we can simply use Optional here and in the lines below?

ijuma commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -995,11 +994,12 @@ class UnifiedLog(@volatile var logStartOffset: Long,
   }
 
-  def latestEpoch: Option[Int] = leaderEpochCache.flatMap(_.latestEpoch)
+  def latestEpoch: Option[Int] = leaderEpochCache.flatMap(_.latestEpoch.asScala).map(Int.unbox(_))
```
Can we use `OptionalInt` here?

satishd replied:
Avoided using Java Optional classes here as all its usages are in Scala classes for now. I prefer making this kind of change as part of moving UnifiedLog to storage module. wdyt?

ijuma replied:
Yes, if all usages are in Scala, it makes sense to keep it for now.

ijuma commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -255,7 +256,7 @@ class RemoteLogManager(rlmConfig: RemoteLogManagerConfig,
 
     // Get the respective epoch in which the starting-offset exists.
-    var maybeEpoch = leaderEpochCache.epochForOffset(startingOffset)
+    var maybeEpoch = leaderEpochCache.epochForOffset(startingOffset).asScala
```
Similarly here, can we just use `Optional` instead of converting to `Option`?

ijuma commented on core/src/test/scala/unit/kafka/cluster/PartitionTest.scala:
```diff
@@ -2650,7 +2647,7 @@ class PartitionTest extends AbstractPartitionTest {
 
     val leaderLog = partition.localLogOrException
-    assertEquals(Some(EpochEntry(leaderEpoch, 0L)), leaderLog.leaderEpochCache.flatMap(_.latestEntry))
+    assertEquals(Some(new EpochEntry(leaderEpoch, 0L)), leaderLog.leaderEpochCache.flatMap(_.latestEntry.asScala))
```
We can change the expected value to be `Optional.of` instead of converting the response with `asScala`.

satishd replied:
`flatMap` has to return `Option`, that is why I converted from `Optional` to `Option`. Another way can be to use match case statements that makes it verbose. I left with Scala Option conversions for now as these will be converted to Java later.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/EpochEntry.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(epoch, startOffset);
```
This results in boxing of the two elements and the allocation of an array. To avoid unexpected performance regressions, can we please stick with an allocation free implementation?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/EpochEntry.java:
```diff
@@ -0,0 +1,51 @@
+        return "EpochEntry{" +
+                "epoch=" + epoch +
+                ", startOffset=" + startOffset +
+                '}';
```
We typically use `(` and `)` versus `{` and `}` in our `toString` implementations.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochCheckpointFile.java:
```diff
@@ -0,0 +1,78 @@
+    private static final Pattern WHITE_SPACES_PATTERN = Pattern.compile("\\s+");
+    private static final int CURRENT_VERSION = 0;
+
+    public static final Formatter FORMATTER = new Formatter();
```
Please have the public field(s) before the private ones.

ijuma reviewed (commented):
Thanks for the PR. Left a few comments I noticed in a first pass. Will take a closer look later.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochCheckpoint.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.List;
+
+public interface LeaderEpochCheckpoint {
```
Looks like this was in the checkpoints package previously. Would it make sense to keep that distinction here? It could be `org.apache.kafka.server.checkpoint.internals` or something along those lines.

ijuma replied:
That said, should this be in the `storage` module at all?

ijuma replied:
It makes sense to me that it should be. But perhaps the package structure for this module should be `org.apache.kafka.server.storage.internals.*` versus the one we've been using for the log layer. Maybe we can use this approach for the checkpoint code and move the log layer all at once at the end.

ijuma replied:
Thinking about this some more, to remain consistent with other packages like `raft`, `metadata` and others, we could skip the `server` and just have `org.apache.kafka.storage.internals.*` for the non api classes.

satishd replied:
+1 on that. I prefer have the package name with `server` but I am fine to omit that to be consistent with the existing convention with other modules. 
I will address it in the next commit of this PR. Let me know if you are fine with the package name as `org.apache.kafka.storage.internals.checkpoint`.

ijuma replied:
I originally also preferred having `server` in the package name, but on further thought, there are advantages in basically having the module name as the package name after `kafka`. It avoids accidental split packages across modules, which can cause problems if we ever decide to use Java 9 modules. And it means that a given server module could eventually be used by clients too if it made sense (without forcing package renames or weird package names).

If we agree on that approach, then the module names would be `org.apache.kafka.storage.internals.checkpoint` and `org.apache.kafka.storage.internals.log` (one annoying aspect of using `log` is that GitHub filters it from their code navigation, but we probably have to live with that for now).

@junrao what are your thoughts?

junrao replied:
Yes, the proposed package names sound good to me.

satishd replied:
I moved checkpoint files to `org.apache.kafka.storage.internals.checkpoint` with the latest commits as mentioned earlier. We can have a separate followup PR for moving existing `log.internals`.

satishd replied:
Filed [KAFKA-14688](https://issues.apache.org/jira/browse/KAFKA-14688) and it is addressed with https://github.com/apache/kafka/pull/13213

satishd committed:
Addressed review comments and resolved unintended merged conflicts.

satishd committed:
Addressed review comments
Avoided Scala versions wherever possible

satishd reviewed (commented):
Thanks @ijuma for your review. Replied with inline comments and addressed with the latest commits.

junrao commented on core/src/test/scala/unit/kafka/server/epoch/LeaderEpochFileCacheTest.scala:
```diff
@@ -243,282 +245,287 @@ class LeaderEpochFileCacheTest {
+    assertEquals(new EpochEntry(1, 7), cache.epochEntries.get(0))
+  }
+
+  def toTuple[K, V](entry: java.util.Map.Entry[K, V]): (K, V) = {
```
Could this be private?

junrao commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -1294,36 +1293,53 @@ class UnifiedLog(@volatile var logStartOffset: Long,
-        val latestEpochOpt = leaderEpochCache.flatMap(_.latestEpoch).map(_.asInstanceOf[Integer])
-        val epochOptional = Optional.ofNullable(latestEpochOpt.orNull)
-        Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, logEndOffset, epochOptional))
+
```
extra newline

junrao reviewed (commented):
@satishd : Thanks for the PR. Just a couple of minor comments.

satishd committed:
Moved LeaderEochCheckpoint and LeaderEpochCheckPointFile to org.apache.kafka.storage.internals.checkpoint package.

ijuma commented on core/src/main/scala/kafka/server/checkpoints/OffsetCheckpointFile.scala:
```diff
@@ -65,9 +65,14 @@ class OffsetCheckpointFile(val file: File, logDirFailureChannel: LogDirFailureCh
     OffsetCheckpointFile.Formatter, logDirFailureChannel, file.getParent)
 
-  def write(offsets: Map[TopicPartition, Long]): Unit = checkpoint.write(offsets)
+  def write(offsets: Map[TopicPartition, Long]): Unit = checkpoint.write(offsets.map{case (k, v)=> Tuple2(k, v)}.toSeq.asJava)
```
The `map` operation is a `no-op` here since we are converting a tuple to a tuple,

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,380 @@
+        }
+    }
+
+    public Optional<Integer> nextEpoch(int epoch) {
```
Can we use `OptionalInt` here?

ijuma commented:
I have a few more comments on this PR, will get back to it soon.

satishd committed:
Addressed reveiw comments

ijuma commented:
@satishd When you have some cycles, please resolve the conflicts on this PR. I think we can focus on merging this next and then #13040.

satishd commented:
@ijuma I rebased with trunk and resolved conflicts. Please review it when you get some time.

ijuma commented on server-common/src/main/java/org/apache/kafka/server/common/CheckpointFile.java:
```diff
@@ -72,7 +71,7 @@ public CheckpointFile(File file,
     }
 
-    public void write(Collection<T> entries) throws IOException {
+    public void write(List<T> entries) throws IOException {
```
Why did we change this?

satishd replied:
This is done to make write and read API consistent like below.

```
public void write(List<T> entries) throws IOException 
public List<T> read() throws IOException 
```

ijuma replied:
I don't think consistency is necessarily a goal in cases like this. The original intent was to accept the minimal interface required, but to return something more specific to avoid unnecessary conversions in both sides.

satishd replied:
Sure, I am fine with the suggestion for now. I updated PR with the changes to keep it like earlier. But this kind of contract looks fragile, we can take a relook at whether we can avoid any conversions and keep the APIs consistent. Filed https://issues.apache.org/jira/browse/KAFKA-14625 to followup later.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/EpochEntry.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public int hashCode() {
+        int result = epoch;
+        result = 31 * result + (int) (startOffset ^ (startOffset >>> 32));
```
Use `Long.hashCode`.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/CheckpointFileWithFailureHandler.java:
```diff
@@ -0,0 +1,63 @@
+public class CheckpointFileWithFailureHandler<T> {
+
+    public final File file;
+    public final int version;
```
`version` was private in the `Scala` implementation, why is it public here?

ijuma reviewed (commented):
Thanks for the update. A few comments below.

satishd committed:
Addressed review comments

satishd committed:
Removed asScala and asJava conversions.

satishd commented:
Thanks @ijuma for your review. Addressed them with inline comments and/or updated commits.

satishd committed:
Addressed review comments

satishd commented:
Thanks @ijuma for the updated comments. Addressed them with inline and update the PR with latest commit.

ijuma commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -249,9 +248,10 @@ class LogSegment private[log] (val log: FileRecords,
+      val maybeCompletedTxn = appendInfo.append(batch, Optional.empty())
       producerStateManager.update(appendInfo)
-      maybeCompletedTxn.foreach { completedTxn =>
+      if (maybeCompletedTxn.isPresent) {
```
Nit: `ifPresent` is the equivalent to `foreach`.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+     * Returns the current Leader Epoch if one exists. This is the latest epoch
+     * which has messages assigned to it.
+     */
+    public Optional<Integer> latestEpoch() {
```
Should this and `previousEpoch` be `OptionalInt`?

ijuma commented on core/src/main/scala/kafka/server/checkpoints/OffsetCheckpointFile.scala:
```diff
@@ -65,9 +65,14 @@ class OffsetCheckpointFile(val file: File, logDirFailureChannel: LogDirFailureCh
     OffsetCheckpointFile.Formatter, logDirFailureChannel, file.getParent)
 
-  def write(offsets: Map[TopicPartition, Long]): Unit = checkpoint.write(offsets)
+  def write(offsets: Map[TopicPartition, Long]): Unit = checkpoint.write(offsets.toSeq.asJava)
```
Is this `toSeq` call still required?

ijuma commented on core/src/main/scala/kafka/server/checkpoints/OffsetCheckpointFile.scala:
```diff
@@ -65,9 +65,14 @@ class OffsetCheckpointFile(val file: File, logDirFailureChannel: LogDirFailureCh
-  def read(): Map[TopicPartition, Long] = checkpoint.read().toMap
+  def read(): Map[TopicPartition, Long] = {
+    val list = checkpoint.read()
+    val result = mutable.Map.empty[TopicPartition, Long]
```
We should size the map appropriately.

ijuma commented on core/src/main/scala/kafka/server/checkpoints/OffsetCheckpointFile.scala:
```diff
@@ -65,9 +65,14 @@ class OffsetCheckpointFile(val file: File, logDirFailureChannel: LogDirFailureCh
+  def read(): Map[TopicPartition, Long] = {
+    val list = checkpoint.read()
+    val result = mutable.Map.empty[TopicPartition, Long]
+    list.forEach { case (tp, offset) => result += tp -> offset }
```
You can avoid the tuple creation by using something like:

`result(tp) = offset`

ijuma commented on core/src/main/scala/kafka/server/checkpoints/OffsetCheckpointFile.scala:
```diff
@@ -65,9 +65,14 @@ class OffsetCheckpointFile(val file: File, logDirFailureChannel: LogDirFailureCh
+    val list = checkpoint.read()
+    val result = mutable.Map.empty[TopicPartition, Long]
+    list.forEach { case (tp, offset) => result += tp -> offset }
+    result.toMap
```
Can we avoid the conversion to an immutable map here? The additional copy doesn't buy us anything.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/CheckpointFileWithFailureHandler.java:
```diff
@@ -0,0 +1,62 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.internals;
```
Why is this not in the checkpoint package?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/EpochEntry.java:
```diff
@@ -0,0 +1,51 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.internals;
```
This was previously under the `epoch` package. Should we do the same, i.e. `storage.internals.epoch`? Same for `LeaderEpochFileCache`.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/checkpoint/LeaderEpochCheckpointFile.java:
```diff
@@ -0,0 +1,83 @@
+        return new File(dir, LEADER_EPOCH_CHECKPOINT_FILENAME);
+    }
+
+    public static class Formatter implements EntryFormatter<EpochEntry> {
```
Does this need to be public?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+        this.checkpoint = checkpoint;
+        LogContext logContext = new LogContext("[LeaderEpochCache " + topicPartition + "] ");
+        log = logContext.logger(LeaderEpochFileCache.class);
+        checkpoint.read().forEach(this::assign);
```
The original code did this under a write lock. Did we establish that it wasn't required?

satishd replied:
Taking write lock of the same instance while initializing the instance does not make sense as the instance is not yet available for access by any other thread while its constructor is still in progress.

ijuma replied:
We are calling a public non-final method from here, so we can publish the instance in an unsafe manner. To make  this change safe, we should make the `assign` method final so it's clear that the instance doesn't escape the constructor.

satishd replied:
`assign(EpochEntry entry)` is already a private method that does not pass this instance outside this class. Am I missing anything here?

ijuma replied:
My bad, we are calling the private `assign` overload., not the public one.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+    }
+
+    private boolean isUpdateNeeded(EpochEntry entry) {
+        Optional<EpochEntry> lastEntry = latestEntry();
```
Nit: can this be inlined?

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/checkpoint/LeaderEpochCheckpoint.java:
```diff
@@ -0,0 +1,29 @@
+
+public interface LeaderEpochCheckpoint {
+
+    void write(Collection<EpochEntry> epochs);
```
Should we change the input of write to a List to indicate that ordering is important and to be more consistent with the return value of read()?

ijuma replied:
@junrao The previous code didn't do that and if we make that change, we have to make additional collection copies in a few places. @satishd filed a separate JIRA to look into this separately (and decide whether it's worth it).

satishd replied:
@junrao This was discussed in an earlier comment [here](https://github.com/apache/kafka/pull/13046#discussion_r1066636586).

junrao reviewed (commented):
@satishd : Thanks for the latest PR. Just one more comment. Also, are the test failures related?

Hangleton commented on storage/src/main/java/org/apache/kafka/storage/internals/checkpoint/LeaderEpochCheckpointFile.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public LeaderEpochCheckpointFile(File file) throws IOException {
+        this(file, null);
```
Do we really want to expose this constructor since nothing prevents an NPE down the line. Should `LogDirFailureChannel` be mocked in unit tests instead?

satishd replied:
Good catch!

Hangleton commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+                if (startOffset == offset) return OptionalInt.of(epoch);
+                if (startOffset > offset) return previousEpoch;
+
+                previousEpoch = OptionalInt.of(epoch);
```
We could use a primitive here to avoid the cost of creating an optional until the value to return is found.

Hangleton commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+    }
+
+    // Visible for testing
+    public List<EpochEntry> epochEntries() {
```
Public but not protected by a lock.

Hangleton commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+        return new ArrayList<>(epochs.values());
+    }
+
+    public void flush() {
```
Ditto.

Hangleton commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+                log.debug("Appended new epoch entry {}. Cache now contains {} entries.", entry, epochs.size());
+            }
+        });
+        flush();
```
Flush only if at least one epoch was assigned?

Hangleton commented on storage/src/main/java/org/apache/kafka/server/log/internals/CheckpointFileWithFailureHandler.java:
```diff
@@ -0,0 +1,62 @@
+        this.file = file;
+        this.logDirFailureChannel = logDirFailureChannel;
+        this.logDir = logDir;
+        checkpointFile = new CheckpointFile<>(file, version, formatter);
```
Curious - the disk failure handler plays an important part in handling I/O failures. What is the reason for not enforcing the same I/O failure path for all types of checkpoint files?

satishd replied:
Each checkpoint file does not need to be treated the same. Whichever checkpoint file needs to treat these errors as disk failure will use the handler.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+    /**
+     * Remove any entries which violate monotonicity prior to appending a new entry
+     */
+    public void maybeTruncateNonMonotonicEntries(EpochEntry newEntry) {
```
This was private before, why is it now public?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+        try {
+            List<EpochEntry> removedEntries = removeFromStart(entry -> entry.startOffset <= startOffset);
+
+            EpochEntry firstBeforeStartOffset;
```
Why do we need this before the `if`?

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+
+                flush();
+
+                log.debug("Cleared entries {} from epoch cache after " + "truncating to end offset {}, leaving {} entries in the cache.", removedEntries, endOffset, epochs.size());
```
No need for string concat here.

ijuma commented on storage/src/main/java/org/apache/kafka/server/log/internals/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,385 @@
+                    }
+                }
+            }
+            log.trace("Processed end offset request for epoch {} and returning epoch {} " + "with end offset {} from epoch cache of size {}}", requestedEpoch, epochAndOffset.getKey(), epochAndOffset.getValue(), epochs.size());
```
No need for the string concat here. Also, since this is a trace log, we may want to have it under a if check (to avoid the array allocation for the parameters, etc.).

ijuma reviewed (commented):
Thanks for the PR. Please see the comments below.

satishd committed:
Addressed review comments
Refactored previousEpoch to return OptionalInt
Refactored latestEpoch to return OptionalInt

satishd committed:
Addressed review comments
Moved CheckpointFileWithFailureHandler to storage.internals.checkpoint and moved LeaderEpochFileCache to storage.internals.epoch

satishd commented:
Thanks @ijuma for your comments. Addressed them with inline replies and/or with the latest commits.

satishd commented:
Thanks @junrao for your review, replied to your comment [inline](https://github.com/apache/kafka/pull/13046#discussion_r1096654123).

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/epoch/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,403 @@
+                log.debug("Appended new epoch entry {}. Cache now contains {} entries.", entry, epochs.size());
+            }
+        });
+        if (entries.size() > 0) flush();
```
This should be `!isEmpty` as some list implementations have a cheaper `isEmpty` than `size()`.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/epoch/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,403 @@
+    public OptionalInt epochForOffset(long offset) {
+        lock.readLock().lock();
+        try {
+            Integer previousEpoch = null;
```
Replacing `OptionalInt` with `Integer` (I believe this was a recent update) doesn't make things better. Both of them box, right? I think the code you had previously was better.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/epoch/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,403 @@
+    }
+
+    public void flush() {
+        lock.writeLock().lock();
```
We added a lock here that wasn't there before. It looks like this is only used within this class. Can we make it private and remove the lock?

satishd replied:
We should have a lock here as this method is accessing epoch entries which can be added/removed from other threads.

ijuma replied:
Yes, but isn't `flush` only called with the lock held? Looking at the usage more closely, this is the case for every usage except `assign`. So, it does look like we have a thread safety bug for the `assign` case. However, the change here has a couple of issues:
1. Holds the lock for the duration of the flush call blocking read operations, which is unnecessary.
2. Introduces new re-entrant lock calls that were not required previously.

ijuma replied:
Also, it's not clear why a `write` lock is required for this operation since we are only reading from `epochs` (the actual checkpoint write has a separate lock).

ijuma replied:
Thinking about it, how about we switch the write lock to a read lock and leave the rest as it is?

satishd replied:
The reason for using write lock is not to allow multiple threads to write concurrently and no updates on epochs. But we can have a constraint on `LeaderEpochCheckpoint#write` implementation to be thread-safe(the implementation already follows this constraint) and use a read lock here. So, as you suggested, taking a read lock here should be sufficient with the mentioned constraint.

ijuma replied:
Let's go with that then as it is least likely to cause a change in behavior from what we had in the Scala code.

ijuma commented on storage/src/main/java/org/apache/kafka/storage/internals/epoch/LeaderEpochFileCache.java:
```diff
@@ -0,0 +1,403 @@
+
+    // Visible for testing
+    public List<EpochEntry> epochEntries() {
+        lock.writeLock().lock();
```
We added a lock that wasn't there before, but this one seems ok since it's only used by tests.

ijuma reviewed (commented):
@satishd Thanks for the updates. Just a few comments below, I think we're very close.

satishd committed:
Addressed review comments

satishd committed:
Use readLock for flush in LeaderEpochFileCache

ijuma reviewed (approved):
LGTM, thanks.

satishd commented:
These test failures seem to be unrelated. I rebased the PR against trunk. I will wait for the test results.

satishd commented:
Test failures do not seem to be related to the PR changes, merging to trunk.
